### DISCUSSION TITLE
Dirichlet boundary condition

### DISCUSSION THREAD

**Discussion by Shihao-Yuan** [SeisSol/SeisSol#616] [Q&A]
Hi,

I am wondering if it is possible to apply a Dirichlet boundary condition to surfaces with a certain tag in SeisSol? 
Thanks!

**Comment by krenzland** [comment]
That depends. Do they depend on time and/or space?

**Reply by Shihao-Yuan** [reply]
For the moment, no, just constant value (e.g. 0).

**Reply by krenzland** [reply]
Then the answer is a strong "maybe". It is possible but not officially supported. You have to create a configuration file that sets the state at the ghost cell.
Like here:
https://github.com/SeisSol/Examples/blob/master/convergence_acoustic_elastic/convergence_ocean/rigid.yaml

This sets the value of the normal velocity in the ghost cell such that the value at the boundary is zero. How you set the values is described in the code itself:
https://github.com/SeisSol/SeisSol/blob/2e2fa4f79f881c905c0a79099c659dd603b1ff95/src/Initializer/ParameterDB.cpp#L362

You can set  the variables   "Tn", "Ts", "Td", "u", "v", "w" at the ghost cell
We read out a affine transformation s.t. val in ghost cell is equal to A * val_inside + b.
Constant terms are named const_{varName}, e.g. const_u
Map terms are named map_{varA}_{varB}, e.g. map_u_v
Mirroring the velocity at the ghost cell would imply the param
map_u_u: -1

E.g. you'd set a zero velocity boundary for elastic material as:
```yaml
!ConstantMap
    map:
      map_u_u: -1.0
      map_v_v: -1.0
      map_w_w: -1.0
```

You also have to mention this in the parameter file like this:
https://github.com/SeisSol/Examples/blob/master/convergence_acoustic_elastic/convergence_ocean/parameters.par#L3

And tag the boundary condition as boundary condition "4".

This should in theory work and should support arbitrary bcs that depend on space but not on time. Unfortunately, this feature is (except for the rigid boundary condition for acoustic) entirely untested and may lead to incorrect results.

**Reply by Shihao-Yuan** [reply]
Thanks Lukas, I will test it.

**Reply by Shihao-Yuan** [reply]
I have tested it in elastic materials and it works well. I am wondering if it could be extended to viscoelastic materials.

**Reply by krenzland** [reply]
Yes, it should work (maybe even straight out of the box?) . They have essentially the same structure.

**Reply by Shihao-Yuan** [reply]
That is also what I expect, but it just stopped at the following step without any error. 
“”“
initializeEasiBoundaryReader with file: rigid.yaml
Mon Jul 18 15:58:01, Info:  .. finished point source initialization.
initializeEasiBoundaryReader with file: rigid.yaml
”“”
I have tested viscoelastic simulations without rigid boundaries and elastic simulations with rigid boundaries. Both work well with the same parameter file (except adding FreqCentral, FreqRatio for the viscoelastic case).

**Reply by krenzland** [reply]
Right, sorry. I've just checked: The feature is not implemented for this equation. You can open up an issue to request the feature. Unfortunately, I do not have time currently to implement it myself.
It should be - in principle - relatively straight-forward. If you want to try implementing it yourself, I could give you a few pointers where to look.
